src/Computation/SymbolicExpression.cs

```diff
@@ -0,0 +1,391 @@
+                    ConstantUnsigned.Create(value.Size, BigInteger.Zero))
+                .Not()
+                .ZeroExtend(Size)
+                .ShiftLeft(offset)
+                .Not();
+
+            var data = value
+                .ZeroExtend(Size)
+                .ShiftLeft(offset);
+
+            return And(mask).Or(data);
+        }
+
+        public IExpression Xor(IExpression expression)
+        {
+            return Binary(expression, (l, r) => new Xor(l, r));
+        }
+
+        public IExpression ZeroExtend(Bits size)
+        {
+            return size > Size
+                ? Unary(v => new ZeroExtend(size, v))
+                : this;
+        }
+
+        private IExpression Unary(Func<IValue, IValue> func)
+        {
+            return new SymbolicExpression(_contextFactory, _collectionFactory,
+                func(Value), Constraints);
+        }
+
+        private IExpression Binary(IExpression y, Func<IValue, IValue, IValue> func)
+        {
+            return Binary((IValueExpression) y, func);
+        }
+
+        private IExpression Binary(IValueExpression y, Func<IValue, IValue, IValue> func)
+        {
+            return new SymbolicExpression(_contextFactory, _collectionFactory,
+                func(Value, y.Value), Constraints.Concat(y.Constraints).ToArray());
+        }
+
+        private IExpression Ternary(IExpression y, IExpression z, Func<IValue, IValue, IValue, IValue> func)
+        {
+            return Ternary((IValueExpression) y, (IValueExpression) z, func);
+        }
+
+        private IExpression Ternary(IValueExpression y, IValueExpression z, Func<IValue, IValue, IValue, IValue> func)
+        {
+            return new SymbolicExpression(_contextFactory, _collectionFactory,
+                func(Value, y.Value, z.Value), Constraints.Concat(y.Constraints.Concat(z.Constraints)).ToArray());
+        }
+
+        private IExpression Evaluate(IPersistentSpace space)
+        {
+            using var model = space.GetModel(Constraints);
+
+            return new ConstantExpression(_contextFactory, _collectionFactory,
+                ConstantUnsigned.Create(Value.Size, model.Evaluate(Value)));
+        }
+
+        private BigInteger AsConstant()
+        {
+            using var handle = _contextFactory.Create();
+
+            return Value is Float && Value.AsFloat(handle.Context).Simplify().IsFPNaN
+                ? Value.Size.GetNan(handle.Context)
+                : AsConstant(Value.AsBitVector(handle.Context).Simplify());
+        }
+
+        private static BigInteger AsConstant(Expr expr)
+        {
+            return expr.IsNumeral
+                ? ((BitVecNum) expr).BigInteger
+                : throw new IrreducibleSymbolicExpressionException();
+        }
+
+        public static SymbolicExpression Create(IContextFactory contextFactory, ICollectionFactory collectionFactory,
+            IValue value, IEnumerable<Func<IExpression, IExpression>>? constraints)
```

Comment:
I'd personally be tempted to make `constraints` non-nullable, given that (and I think this applies here) an empty enumerable serves the same purpose.